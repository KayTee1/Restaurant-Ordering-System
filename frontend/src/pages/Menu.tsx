import React, { useState, useCallback, useEffect } from "react";
import axios from "axios";

import { Container, Row, Col } from "react-bootstrap";
import PacmanLoader from "react-spinners/PacmanLoader";
import MenuItemCard from "../components/MenuItemCard";

import Dish from "../types/Dish";

const Menu: React.FC = () => {
  const [dishes, setDishes] = useState<Dish[]>([]);
  const [isLoading, setIsLoading] = useState<boolean>(true);
  const [isError, setIsError] = useState<boolean>(false);

  const fetchDishes = useCallback(async () => {
    try {
      const dishesUrl = "http://localhost:5000/api/dishes";
      const response = await axios.get<Dish[]>(dishesUrl);
      const data = response.data;

      setDishes(data);
    } catch (err) {
      console.error(err);
      setIsError(true);
    }
    setIsLoading(false);
  }, []);

  useEffect(() => {
    fetchDishes();
  }, [fetchDishes]);

  let renderContent;

  isError
    ? (renderContent = <p>Something Went Wrong! :(</p>)
    : (renderContent = (
        <Row md={2} xs={1} className="g-3">
          {dishes.map((dish: Dish, index: number) => (
            <Col key={index}>
              <MenuItemCard key={index} dish={dish} />
            </Col>
          ))}
        </Row>
      ));

  return (
    <>
      <Container>
        <h1 className="mb-3">Menu</h1>
        {isLoading ? (
          <PacmanLoader
            color="#295cba"
            loading={isLoading}
            size={50}
            aria-label="Loading Spinner"
            data-testid="loader"
          />
        ) : (
          renderContent
        )}
      </Container>
    </>
  );
};

export default Menu;
