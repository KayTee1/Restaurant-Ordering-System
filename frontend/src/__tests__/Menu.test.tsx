import { expect, it, describe } from "vitest";
import { MemoryRouter } from "react-router-dom";
import { createRenderer } from "react-test-renderer/shallow";

import Menu from "../pages/Menu";

const renderer = createRenderer();

const defaultComponent = (
  <MemoryRouter initialEntries={["/"]}>
    <Menu />
  </MemoryRouter>
);

// Tests
describe("On the Menu page", () => {
  it("should render and match the snapshot", () => {
    renderer.render(defaultComponent);
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });

  it("should render h1", () => {
    renderer.render(defaultComponent);
    const h1 = renderer.getRenderOutput();
    expect(h1).toBeTruthy();
  });
});
