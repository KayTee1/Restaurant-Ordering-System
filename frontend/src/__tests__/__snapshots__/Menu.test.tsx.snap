// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`On the Menu page > should render and match the snapshot 1`] = `
<Router
  location={
    {
      "hash": "",
      "key": "default",
      "pathname": "/",
      "search": "",
      "state": null,
    }
  }
  navigationType="POP"
  navigator={
    {
      "action": "POP",
      "createHref": [Function],
      "createURL": [Function],
      "encodeLocation": [Function],
      "go": [Function],
      "index": 0,
      "listen": [Function],
      "location": {
        "hash": "",
        "key": "default",
        "pathname": "/",
        "search": "",
        "state": null,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Menu />
</Router>
`;
